ShelterModule/Services/Implementations/Announcements/AnnouncementQuery.cs

```diff
@@ -0,0 +1,57 @@
+﻿using Database;
+using Microsoft.EntityFrameworkCore;
+using ShelterModule.Models.Announcements;
+using ShelterModule.Models.Pets;
+using ShelterModule.Models.Shelters;
+using ShelterModule.Services.Interfaces.Announcements;
+
+namespace ShelterModule.Services.Implementations.Announcements;
+
+public class AnnouncementQuery : IAnnouncementQuery
+{
+    private readonly PetShareDbContext _context;
+
+    public AnnouncementQuery(PetShareDbContext context)
+    {
+        _context = context;
+    }
+
+    public async Task<IReadOnlyList<Announcement>> GetAllFilteredAsync(GetAllAnnouncementsFilteredQuery query,
+        CancellationToken token = default)
+    {
+        var filteredAnnouncements =
+            (await _context.Announcements.Include(x => x.Author).Include(x => x.Pet).ToListAsync(token)).
+            Select(a => new
+            {
+                Announcement = Announcement.FromEntity(a),
+                Pet = new { Pet = Pet.FromEntity(a.Pet), Shelter = Shelter.FromEntity(a.Pet.Shelter) },
+                Author = Shelter.FromEntity(a.Author)
+            });
+
+        if (query.Species is not null)
+            filteredAnnouncements = filteredAnnouncements.Where(a => query.Species.Contains(a.Pet.Pet.Species));
+        if (query.Breeds is not null)
+            filteredAnnouncements = filteredAnnouncements.Where(a => query.Breeds.Contains(a.Pet.Pet.Breed));
+        if (query.Cities is not null)
+            filteredAnnouncements = filteredAnnouncements.Where(a => query.Cities.Contains(a.Pet.Shelter.Address.City));
+        if (query.MinAge is not null)
+            filteredAnnouncements =
+                filteredAnnouncements.Where(a => DateTime.Now.Year - a.Pet.Pet.Birthday.Year >= query.MinAge);
+        if (query.MaxAge is not null)
```

Comment:
We can have all those `Where()` clauses earlier, before calling `ToListAsync()`. Right now we pull all announcements from db, and then filter them in memory. If we have `Where()`s earlier, they will be translated to SQL, so we will only pull those announcements form db, that we actually need.